File: kimbits.cpp
/*
ID: totheso1
LANG: C++14
TASK: kimbits
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <limits>
#include <vector>

using namespace std;

static ifstream fin("kimbits.in");
static ofstream fout("kimbits.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static const auto LENGTH = fin_get<int>();
static const auto MAX_BIT_COUNT = fin_get<int>();
static const auto RANK = fin_get<unsigned>();
static vector<vector<unsigned>> count_collection(
    LENGTH + 1, vector<unsigned>(MAX_BIT_COUNT + 1, 1));

static void initialize() {
  for (int length = 1; length <= LENGTH; ++length) {
    for (int max_bit_count = 1; max_bit_count <= MAX_BIT_COUNT;
         ++max_bit_count) {
      count_collection[length][max_bit_count] =
          count_collection[length - 1][max_bit_count] +
          count_collection[length - 1][max_bit_count - 1];
    }
  }
}

static void print_bits(int length, int max_bit_count, unsigned rank) {
  while (length > 0) {
    const auto tmp = count_collection[length - 1][max_bit_count];
    if (rank <= tmp) {
      fout << 0;
    } else {
      fout << 1;
      --max_bit_count;
      rank -= tmp;
    }
    --length;
  }
  fout << endl;
}

int main() {
  initialize();
  print_bits(LENGTH, MAX_BIT_COUNT, RANK);
}

File: kimbits.py
"""
ID: totheso1
LANG: PYTHON3
TASK: kimbits
"""

from typing import Any

fin = open("kimbits.in", "r")
fout = open("kimbits.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
LENGTH, MAX_BIT_COUNT, RANK = map(int, fin.readline().split())
count_collection = [[1] * (MAX_BIT_COUNT + 1) for _ in range(LENGTH + 1)]


def initialize() -> None:
    for length in range(1, LENGTH + 1):
        for max_bit_count in range(1, MAX_BIT_COUNT + 1):
            count_collection[length][max_bit_count] = (
                count_collection[length - 1][max_bit_count]
                + count_collection[length - 1][max_bit_count - 1]
            )


def print_bits(length: int, max_bit_count: int, rank: int) -> None:
    while length > 0:
        tmp = count_collection[length - 1][max_bit_count]
        if rank <= tmp:
            fprint(0, end="")
        else:
            fprint(1, end="")
            max_bit_count -= 1
            rank -= tmp
        length -= 1
    fprint()


initialize()
print_bits(LENGTH, MAX_BIT_COUNT, RANK)
# Code end
